File: tetris_randomizer.py
import random

class TetrisBagRandomizer:
    def __init__(self):
        self.tetrominos = ['I', 'O', 'T', 'J', 'L', 'S', 'Z']
        self.current_bag = []
        self._generate_new_bag()
    
    def _generate_new_bag(self):
        """Tạo một bag mới bằng cách xáo trộn ngẫu nhiên 7 khối tetromino."""
        self.current_bag = self.tetrominos.copy()
        random.shuffle(self.current_bag)
    
    def next_piece(self):
        """Lấy ra khối tetromino tiếp theo từ bag hiện tại."""
        if not self.current_bag:
            self._generate_new_bag()
        return self.current_bag.pop()
    
    def peek_next_pieces(self, count=1):
        """Xem trước các khối tiếp theo mà không lấy ra khỏi bag."""
        pieces = []
        temp_bag = self.current_bag.copy()
        
        for _ in range(count):
            if not temp_bag:
                # Tạo bag mới tạm thời để xem trước
                temp_bag = self.tetrominos.copy()
                random.shuffle(temp_bag)
            pieces.append(temp_bag.pop())
            
        return pieces

    def get_next_bag(self):
        """Trả về một bag mới gồm 7 khối xáo trộn."""
        bag = self.tetrominos.copy()
        random.shuffle(bag)
        return bag


# Ví dụ sử dụng
if __name__ == "__main__":
    randomizer = TetrisBagRandomizer()
    
    print("Lấy từng mảnh:")
    for _ in range(10):
        print(randomizer.next_piece(), end=" ")
    
    print("\n\nXem trước 3 mảnh tiếp theo:")
    print(randomizer.peek_next_pieces(3))
    
    print("\nLấy một bag mới hoàn chỉnh:")
    print(randomizer.get_next_bag()) 

File: tetris_randomizer.js
/**
 * Hệ thống 7-bag randomizer cho Tetris
 * Mỗi bag chứa 7 khối Tetris (I, O, T, J, L, S, Z) được xáo trộn ngẫu nhiên
 */
class TetrisBagRandomizer {
  constructor() {
    this.tetrominos = ['I', 'O', 'T', 'J', 'L', 'S', 'Z'];
    this.currentBag = [];
    this.generateNewBag();
  }

  /**
   * Tạo một bag mới bằng cách xáo trộn 7 khối tetromino
   * @private
   */
  generateNewBag() {
    // Tạo bản sao của mảng tetrominos
    this.currentBag = [...this.tetrominos];
    
    // Thuật toán Fisher-Yates để xáo trộn mảng
    for (let i = this.currentBag.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [this.currentBag[i], this.currentBag[j]] = [this.currentBag[j], this.currentBag[i]];
    }
  }

  /**
   * Lấy khối tetromino tiếp theo từ bag
   * @returns {string} Khối tetromino tiếp theo
   */
  nextPiece() {
    // Nếu bag trống, tạo bag mới
    if (this.currentBag.length === 0) {
      this.generateNewBag();
    }
    
    // Lấy và trả về khối cuối cùng trong bag
    return this.currentBag.pop();
  }

  /**
   * Xem trước các khối tiếp theo mà không lấy ra khỏi bag
   * @param {number} count - Số lượng khối muốn xem trước
   * @returns {Array} Mảng chứa các khối tiếp theo
   */
  peekNextPieces(count = 1) {
    const pieces = [];
    const tempBag = [...this.currentBag];
    let tempFutureBag = [];
    
    for (let i = 0; i < count; i++) {
      if (tempBag.length === 0) {
        tempFutureBag = [...this.tetrominos];
        // Xáo trộn bag tạm thời
        for (let i = tempFutureBag.length - 1; i > 0; i--) {
          const j = Math.floor(Math.random() * (i + 1));
          [tempFutureBag[i], tempFutureBag[j]] = [tempFutureBag[j], tempFutureBag[i]];
        }
        pieces.push(tempFutureBag.pop());
      } else {
        pieces.push(tempBag.pop());
      }
    }
    
    return pieces;
  }
}

// Export module nếu đang chạy trong Node.js
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { TetrisBagRandomizer };
} 